### ISSUE TITLE
Refactor ArchiveResult filesystem reads and writes to go through Django file storage backend

### ISSUE THREAD

**Issue by pirate** [ArchiveBox/ArchiveBox#902]
Instead of this:
```python3
class ArchiveResult:
    path = field.CharField(...)

ArchiveResult(path='./archive/warc/somefile.warc.gz')
```
We should be doing this:
```python3
class ArchiveResult:
    path = field.FileField(...)

ArchiveResult(path=Path('./archive/warc/somefile.warc.gz'))
```

`settings.py`:
```python3
MEDIA_URL = 'archive'
MEDIA_ROOT = 'archive'

DEFAULT_FILE_STORAGE = 'django.core.files.storage.FileSystemStorage'
# or
DEFAULT_FILE_STORAGE = 'storages.backends.s3boto3.S3Boto3Storage'   # or DigitalOcean, SFTP, etc.
```

https://django-storages.readthedocs.io/en/latest/
https://docs.djangoproject.com/en/4.0/topics/files/

related: https://github.com/ArchiveBox/ArchiveBox/issues/788

**Comment by mAAdhaTTah** [comment]
How do you imagine this interacting with the extractors? Will they write via Django storage?

**Comment by pirate** [comment]
Yeah, I have a big re-archictecturing plan to move to plug-in-style hooks system that allow components to modify any archivebox behavior, including filesystem calls, extractors, parsing, etc.

**Comment by pirate** [comment]
Actually I'm curious what you think of the new architecture, do you find this intuitive @mAAdhaTTah? https://gist.github.com/pirate/7193ab54557b051aa1e3a83191b69793

**Comment by caj-larsson** [comment]
Interesting, I've been thinking about building a plugable extractor but if you are already working on this on a whole application scale I could help out with this instead.

I don't quite understand how available hooks in the internals will be defined. 
Is it correct that, because of the deep merge behavior, the last plugin is called first?

**Comment by opentyler** [comment]
This seems like a great solution to #940. Being able to use an external storage would be incredible. 

What do you think are the pain points for integrating this solution? Would love to be able to help

**Comment by pirate** [comment]
FYI all I just created a new Wiki page covering how to set up ArchiveBox with a remote filesystem: https://github.com/ArchiveBox/ArchiveBox/wiki/Setting-Up-Storage

I confirmed it works myself with Amazon S3, Backblaze B2, SFTP, SMB, NFS. It should also work with Google Drive, OneDrive, DropBox, and all the other platforms that RClone supports.
